From patchwork Sat Jul 21 18:56:57 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Eugene Korenevsky <ekorenevsky@gmail.com>
X-Patchwork-Id: 10539045
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id CA490A6A5
	for <patchwork-kvm@patchwork.kernel.org>;
 Sun, 22 Jul 2018 00:57:34 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 42A2628AE7
	for <patchwork-kvm@patchwork.kernel.org>;
 Sat, 21 Jul 2018 18:57:06 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 34DAC28AFC; Sat, 21 Jul 2018 18:57:06 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id D346528AE7
	for <patchwork-kvm@patchwork.kernel.org>;
 Sat, 21 Jul 2018 18:57:05 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728314AbeGUTuq (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Sat, 21 Jul 2018 15:50:46 -0400
Received: from mail-wr1-f67.google.com ([209.85.221.67]:33863 "EHLO
        mail-wr1-f67.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727973AbeGUTuq (ORCPT <rfc822;kvm@vger.kernel.org>);
        Sat, 21 Jul 2018 15:50:46 -0400
Received: by mail-wr1-f67.google.com with SMTP id c13-v6so14119833wrt.1
        for <kvm@vger.kernel.org>; Sat, 21 Jul 2018 11:57:02 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=date:from:to:subject:message-id:mail-followup-to:mime-version
         :content-disposition:user-agent;
        bh=r1Uo+/4yO7oiYlZb3W6y4Py3Negf/6/UcTZ+8MG6YS8=;
        b=fFRjIY5qMjFANlEsbLh0WJQVcKnJpjMpXEAVKr3fpi0DgpGOTHrjaUtTV6CcXAPuRh
         T2d2mtdItCjUPr7TcVN60KpfGvu8AstBogTKymY9CKEVo4mGHE8kvb5csh3hFRQ0TEBK
         c+z3SHHVDrcjoUYfZxmtX7dj2/GL1YDGiP7LPJi9eoPSGYUIpd8dHg/teS11eAy0Jlvp
         kt2Oe5Kc7XyyBtqg7VsEgmGIwdrPYY8VLZZ1CNR43ydKPPMwgucSbIuUDshYd0swvJqN
         mVMwYfbPYUehnKq8daLHMnHzBJmAESKtY7QsamSn1Dwiqr6BJuE6yYYqsB5yH5JCjK8a
         J0rw==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:from:to:subject:message-id:mail-followup-to
         :mime-version:content-disposition:user-agent;
        bh=r1Uo+/4yO7oiYlZb3W6y4Py3Negf/6/UcTZ+8MG6YS8=;
        b=N9Q+ojKjX++QTCkVv9TnxWQsiHzdJvpZY4l+Vzhl/Hyo9O9KmcToUB1twh9x5ESbB/
         3u+t/wk9NEi0ueV2JTTCr4o8hiCr4Tw+AHkOVM88MEQ76w6wZMpfVV1dTZahrXphBh88
         eTduxeRyxcO2vpWrZZftxm3MCabEiABE2d77iAK3VgqGZs1sq747OUnL9AFkx/JrFL7P
         dZUJyJ3E0sP6aMNf7JZJnAJBBnMGo1UUVEtmrpJWBTaCrk8fUDaZ5jF6NZj2VU1BJoj0
         yKh/ypg1T5WCjuGjYzISr2AOjbrk6A6knWZqqelc4JfZTt4iAp9UIEgIi+4OicPYST0v
         3F5Q==
X-Gm-Message-State: AOUpUlF4SAFlTrFd77c6Ykv51RhxrS5Gqet4HL26/ktRBKtUTHQJWU7n
        QpPfb2opcMx7T1rtwsvNzsPTZbpi
X-Google-Smtp-Source: 
 AAOMgpe509iPAYd0F6mU4WlFU+knkiMiA6Boqfizdg+gVjvcynnahIf5BozmzYiaEpPuw5i7ZAEE+g==
X-Received: by 2002:adf:b112:: with SMTP id
 l18-v6mr4756847wra.101.1532199422178;
        Sat, 21 Jul 2018 11:57:02 -0700 (PDT)
Received: from vnote ([95.72.113.226])
        by smtp.gmail.com with ESMTPSA id
 j6-v6sm8217610wrs.91.2018.07.21.11.57.01
        (version=TLS1_2 cipher=ECDHE-RSA-CHACHA20-POLY1305 bits=256/256);
        Sat, 21 Jul 2018 11:57:01 -0700 (PDT)
Date: Sat, 21 Jul 2018 21:56:57 +0300
From: Eugene Korenevsky <ekorenevsky@gmail.com>
To: kvm@vger.kernel.org, Paolo Bonzini <pbonzini@redhat.com>,
        Jan Kiszka <jan.kiszka@web.de>
Subject: [kvm-unit-tests PATCH 1/2] x86: io: fix 8250 serial controller
 initialization
Message-ID: <20180721185657.GA21275@vnote>
Mail-Followup-To: Eugene Korenevsky <ekorenevsky@gmail.com>,
        kvm@vger.kernel.org, Paolo Bonzini <pbonzini@redhat.com>,
        Jan Kiszka <jan.kiszka@web.de>
MIME-Version: 1.0
Content-Type: text/plain; charset=us-ascii
Content-Disposition: inline
User-Agent: Mutt/1.10.1 (2018-07-13)
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

These changes make serial controller working at real hardware and BOCHS:
- Disable interrupts
- Set up LCR, FCR, MCR properly

Signed-off-by: Eugene Korenevsky <ekorenevsky@gmail.com>
---
 lib/x86/io.c | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/lib/x86/io.c b/lib/x86/io.c
index 7e1c16d..f697161 100644
--- a/lib/x86/io.c
+++ b/lib/x86/io.c
@@ -39,6 +39,15 @@ static void serial_init(void)
         lcr = inb(serial_iobase + 0x03);
         lcr &= ~0x80;
         outb(lcr, serial_iobase + 0x03);
+
+        /* IER: disable interrupts */
+        outb(0x00, serial_iobase + 0x01);
+        /* LCR: 8 bits, no parity, one stop bit */
+        outb(0x03, serial_iobase + 0x03);
+        /* FCR: disable FIFO queues */
+        outb(0x00, serial_iobase + 0x02);
+        /* MCR: RTS, DTR on */
+        outb(0x03, serial_iobase + 0x04);
 }
 
 static void print_serial(const char *buf)

From patchwork Sat Jul 21 18:59:15 2018
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: Eugene Korenevsky <ekorenevsky@gmail.com>
X-Patchwork-Id: 10539029
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id B9575A698
	for <patchwork-kvm@patchwork.kernel.org>;
 Sun, 22 Jul 2018 00:57:34 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 59DF028C6A
	for <patchwork-kvm@patchwork.kernel.org>;
 Sat, 21 Jul 2018 18:59:27 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 4B8C228C6E; Sat, 21 Jul 2018 18:59:27 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-8.0 required=2.0 tests=BAYES_00,DKIM_SIGNED,
	DKIM_VALID,DKIM_VALID_AU,FREEMAIL_FROM,MAILING_LIST_MULTI,RCVD_IN_DNSWL_HI
	autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id DA8B328C6A
	for <patchwork-kvm@patchwork.kernel.org>;
 Sat, 21 Jul 2018 18:59:26 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1728028AbeGUTxI (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Sat, 21 Jul 2018 15:53:08 -0400
Received: from mail-wm0-f65.google.com ([74.125.82.65]:35900 "EHLO
        mail-wm0-f65.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
        with ESMTP id S1727997AbeGUTxI (ORCPT <rfc822;kvm@vger.kernel.org>);
        Sat, 21 Jul 2018 15:53:08 -0400
Received: by mail-wm0-f65.google.com with SMTP id s14-v6so12754906wmc.1
        for <kvm@vger.kernel.org>; Sat, 21 Jul 2018 11:59:24 -0700 (PDT)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=gmail.com; s=20161025;
        h=date:from:to:subject:message-id:mail-followup-to:mime-version
         :content-disposition:user-agent;
        bh=bWapW6F1zAUfiaRqHnS7GuFYBHqMLZw6IXFezMxs9Iw=;
        b=n0CAo9ek5OI+rqg3wUqkzwhnVVzMy68Ena0n25+Aey49zgO4OzdhRrZRpNeWrmu5ZB
         QqTGd+pSiAdlpgiN4eZR4x3QtEFucHnRfBNlth8FuJM/q5KEeOGRJjPV4S3ZE25hZjWS
         tXIOOt0wdng2wXvoV7w3XjyC6Eaaii92p9qaEarqkXerV9TLbk9hIIWZZKXHcx04GxQv
         +3k9NYGc0DnzU1mYFcuuvvWc+aL3zryAcR7Xga4B60GWoyxXTjW8USOrJ+shWlhD4v9D
         E/3iSY4uMBSMye9+VaTi/H5Ausa+eOukPD+Ll3E9Yin/LbHtJkBxXB7oD+CXe+jm5U+a
         8m1A==
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
        d=1e100.net; s=20161025;
        h=x-gm-message-state:date:from:to:subject:message-id:mail-followup-to
         :mime-version:content-disposition:user-agent;
        bh=bWapW6F1zAUfiaRqHnS7GuFYBHqMLZw6IXFezMxs9Iw=;
        b=hh6QPjjcpG6MbnZWSnbKX27d763MNVi0yTN7Z3xP8Fbu3WTjE+m5fDBKjLphzU+l06
         TkF6dfsNl4gi4XRDspBC7AaSsebI71yMtu47dr0KkKZ515GWKy+vyWfBYOZi+mtxiGZd
         vKicYlc2t+8RqbqsFTF0dsPuHYex1FuRSLtrVeUXl6nhpyUbqZp5/LMraCPEC7PuCRdE
         +H2ChSl3zDj99HdktpJYL+HVDzk521j3b2NPs+jo97lxQf2KXDsith0CZXr4r4eRjdAD
         H2gZgzr+dURPrwAn+Mhm9UlmGyxJnV/k7uPcrdG1n5XATDVG2gk8D3ptIyX+a1Nb/DZQ
         e7Bw==
X-Gm-Message-State: AOUpUlGL7xtGdzu8uP9AdeWK0BFZ/0HdvTV74SFgY1AOcC/UEJMmlWvQ
        ykdpp8lrr3h2OGDEUqIDSIlO0e0y
X-Google-Smtp-Source: 
 AAOMgpfvtTEwQ1Sf1KeUND7wxvTfnE7JpuUMmVF00m8XP4YQOUOjS9Istrr54QiZG7LZiIBmJaCjwQ==
X-Received: by 2002:a1c:888e:: with SMTP id
 k136-v6mr4281967wmd.6.1532199563359;
        Sat, 21 Jul 2018 11:59:23 -0700 (PDT)
Received: from vnote ([95.72.113.226])
        by smtp.gmail.com with ESMTPSA id
 f6-v6sm3609354wrp.30.2018.07.21.11.59.22
        (version=TLS1_2 cipher=ECDHE-RSA-CHACHA20-POLY1305 bits=256/256);
        Sat, 21 Jul 2018 11:59:22 -0700 (PDT)
Date: Sat, 21 Jul 2018 21:59:15 +0300
From: Eugene Korenevsky <ekorenevsky@gmail.com>
To: kvm@vger.kernel.org, Paolo Bonzini <pbonzini@redhat.com>,
        Jan Kiszka <jan.kiszka@web.de>
Subject: [kvm-unit-tests PATCH 2/2] x86: io: force carriage return on \n put
 to serial port
Message-ID: <20180721185915.GA21451@vnote>
Mail-Followup-To: Eugene Korenevsky <ekorenevsky@gmail.com>,
        kvm@vger.kernel.org, Paolo Bonzini <pbonzini@redhat.com>,
        Jan Kiszka <jan.kiszka@web.de>
MIME-Version: 1.0
Content-Type: text/plain; charset=us-ascii
Content-Disposition: inline
User-Agent: Mutt/1.10.1 (2018-07-13)
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

When '\n' (line feed) is put to serial port, put '\r' (carriage return) as
well. This makes the serial port output to look properly (without stairs) at
real hardware. Also the output looks well at QEMU/KVM and BOCHS emulators.

Signed-off-by: Eugene Korenevsky <ekorenevsky@gmail.com>
---
 lib/x86/io.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/lib/x86/io.c b/lib/x86/io.c
index f697161..edac5bc 100644
--- a/lib/x86/io.c
+++ b/lib/x86/io.c
@@ -22,6 +22,14 @@ static void serial_outb(char ch)
         outb(ch, serial_iobase + 0x00);
 }
 
+static void serial_put(char ch)
+{
+        /* Force carriage return to be performed on \n */
+        if (ch == '\n')
+                serial_outb('\r');
+        serial_outb(ch);
+}
+
 static void serial_init(void)
 {
         u8 lcr;
@@ -61,7 +69,7 @@ static void print_serial(const char *buf)
         }
 
         for (i = 0; i < len; i++) {
-            serial_outb(buf[i]);
+            serial_put(buf[i]);
         }
 #else
         asm volatile ("rep/outsb" : "+S"(buf), "+c"(len) : "d"(0xf1));
